src/Cooked/PrintToBuiltinString.hs

```diff
@@ -0,0 +1,357 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# HLINT ignore "Use camelCase" #-}
+{-# HLINT ignore "Use foldr" #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
+
+-- | Print all the types that occur on the 'TxInfo' to 'BuiltinString'. This is
+-- useful for debugging of validators. You probably do not want to use this in
+-- production code, as many of the functions in this module are wildly
+-- inefficient due to limitations of the 'BuiltinString' type.
+module Cooked.PrintToBuiltinString (PrintBS (..), printBSs, app_prec) where
+
+import Plutus.V2.Ledger.Api
+import qualified PlutusTx.AssocMap as PlMap
+import PlutusTx.Builtins
+import PlutusTx.Prelude
+
+-- | analogue of Haskell's 'Show' class for use in Plutus scripts.
+class PrintBS a where
+  -- | analogue of 'show'
+  {-# INLINEABLE printBS #-}
+  printBS :: a -> BuiltinString
+  printBS x = printBSsPrec 0 x ""
+
+  -- | analogue of 'showsPrec'
+  {-# INLINEABLE printBSsPrec #-}
+  printBSsPrec :: Integer -> a -> BuiltinString -> BuiltinString
+  printBSsPrec _ x s = printBS x <> s
+
+-- | analogue of 'shows'
+{-# INLINEABLE printBSs #-}
+printBSs :: PrintBS a => a -> BuiltinString -> BuiltinString
+printBSs = printBSsPrec 0
+
+-- | Precedence of function applications
+{-# INLINEABLE app_prec #-}
+app_prec :: Integer
+app_prec = 10
+
+{-# INLINEABLE literal #-}
+literal :: BuiltinString -> BuiltinString -> BuiltinString
+literal = (<>)
+
+{-# INLINEABLE cat #-}
+cat :: [BuiltinString] -> BuiltinString -> BuiltinString
+cat [] = id -- we cannot use foldr here for some reason with the plutus compiler
+cat (x : xs) = literal x . cat xs
```

Comment:
Perhaps this can be defined as 
```haskell
cat xs x = foldr (<>) x xs
```
Otherwise, what's the problem that the plutus compiler reports?